### ISSUE TITLE
MaskRCNN Unable to pass a batch and slow inference

### ISSUE THREAD

**Issue by rbrigden** [pytorch/vision#1054]
The new torchvision maskRCNN model does not take in a tensor of form (B,C,H,W) where B is Batch Size but instead it takes a list of tensors. 

Furthermore, inference on a list with 5 tensors on the maskRCNN model takes about 0.5 seconds on a 1080Ti which is a lot slower than what is published on the pytorch blog 0.0903. The inference also seems to scale up linearly as the list of tensors passed in grows. 

Is there a way to not have this linear increase in inference speed when using the model? 

Thanks

**Comment by fmassa** [comment]
Hi @rbrigden,

Those are great questions. Let me address them separately:

> The new torchvision maskRCNN model does not take in a tensor of form (B,C,H,W) where B is Batch Size but instead it takes a list of tensors.

The model takes a list of images because it's strictly more general than passing a batch, as it support images of different sizes.
Internally, all images are padded to be the same size and are batched, see https://github.com/pytorch/vision/blob/15e24bdfe0ba04d728afcff005ff2e01f1da9b33/torchvision/models/detection/transform.py#L88-L103

If you have a batched tensor, you can instead just pass `tensor.unbind(0)` to return a list of tensors.

> The inference also seems to scale up linearly as the list of tensors passed in grows.

The model is performing most of the computations in a batched fashion. Indeed, most of the heavy operations are running on batches, and the few for loops that are batch-size dependent are not meaningfully slow compared to the other (bigger) ops, like convolutions (which are batched).

The problem is that the images that are fed to `MaskRCNN` are pretty big (minimum size is 800 by default). This is enough to occupy the GPU already, so having them batched does not necessarily bring any speed benefits.

Here is an example of the first convolution in Mask R-CNN, which a representative size, on a P100 GPU:
```python
In [1]: import torch

In [2]: a = torch.rand(2, 3, 800, 800).cuda()

In [3]: m = torch.nn.Conv2d(3, 64, 7, 2, 3).cuda()

In [4]: o = m(a)

In [5]: o.shape
Out[5]: torch.Size([2, 64, 400, 400])

In [6]: %timeit m(a); torch.cuda.synchronize()
1.37 ms ± 2.74 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [7]: a = torch.rand(4, 3, 800, 800).cuda()

In [8]: o = m(a)

In [9]: %timeit m(a); torch.cuda.synchronize()
2.68 ms ± 10.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [10]: a = torch.rand(6, 3, 800, 800).cuda()

In [11]: o = m(a)

In [12]: %timeit m(a); torch.cuda.synchronize()
4.02 ms ± 22 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [13]: a = torch.rand(8, 3, 800, 800).cuda()

In [14]: o = m(a)

In [15]: %timeit m(a); torch.cuda.synchronize()
5.34 ms ± 20.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

As you can see, there is an almost linear scaling with the batch size, because the original tensor size is already so big.

> Furthermore, inference on a list with 5 tensors on the maskRCNN model takes about 0.5 seconds on a 1080Ti which is a lot slower than what is published on the pytorch blog 0.0903.

The results that I have presented in the blog correspond to a batch of size 1, which is 0.0903 s on a V100.
So 0.5 seconds for 5 images in the batch sounds reasonable to me on a 1080Ti, given the above reasoning.

---

To summarize, the linear increase is not due to for loops on the batch size, but instead because the GPUs are already saturated by the large inputs, so batching does not increase utilization significantly -- and thus the linear increase.

**Comment by rbrigden** [comment]
Thanks a lot for the detailed response! Is there a reason why the images are padded to 800 minimum size?

**Comment by fmassa** [comment]
@rbrigden the images are resized to a minimum of 800 pixels in size so that it can detect smaller objects better. It is how it has been originally done in the Mask R-CNN paper.

The zero padding is added afterwards to make the images be of the same size.

**Comment by elepherai** [comment]
@fmassa Hi, do you have any idea to make full use of GPU memory and speed up inference?

**Comment by fmassa** [comment]
@elepherai You can get some small speedups if using larger batch size, but as I mentioned in https://github.com/pytorch/vision/issues/1054#issuecomment-505791311, the original images are already pretty big, so increasing batch size won't improve much the inference speed because the GPU is already being fully-utilized

**Comment by elepherai** [comment]
@fmassa But even the min size is 800, there are still more than half of memories left in 12GB GPU. So I’m not clear about 

>  the GPU is already being fully-utilized

Could you explain that? Thanks~

**Comment by elepherai** [comment]
BTW, I have tested maskrcnn-benchmark batch inference lately and found that the time of rpn increased linearly while backbones were almost identical.

Batch 3 vs. Batch 1

![PNG图像](https://user-images.githubusercontent.com/43477405/67708022-76af9400-f9f6-11e9-9c95-82ee6d7d6b32.png)

**Comment by fmassa** [comment]
@elepherai can you try the detection models in torchvision instead? I've optimized a few things (by fusing a few operations) which should make inference on torchvision faster than in maskrcnn-benchmark.
Plus, if you paste a minimal script which reproduces the behavior it would make it much easier to debug.

**Comment by elepherai** [comment]
@fmassa Thanks, I will try torchvision first.
